Enhancing Deployment Documentation (#1607)
Kludex/uvicorn: docs/deployment/getting-started.md

# Deployment

Server deployment is a complex area, that will depend on what kind of service you're deploying Uvicorn onto.

### TL;DR
As a general rule, you probably want to:

* Run `uvicorn --reload` from the command line for local development.
* Run `gunicorn -k uvicorn.workers.UvicornWorker` for production (non containerized).
* Run behind Nginx for self-hosted deployments.
* Finally, run everything behind a CDN for caching support, and serious DDOS protection.

## Quick Start

> In the Quick Start section, we will cover how to start serving your application. The application mentioned here is any ASGI compatible application. For instance if an app that was built using a framework like `Starlette` or `FastAPI`.

Reviewer: ```suggestion
This section is meant to get you started serving an ASGI application as quick as possible. ASGI applications are applications that adhere to the [ASGI Specification], which is the contract between the ASGI server (Uvicorn) and the ASGI application. Starlette  is just one ASGI framework amongst many, see [awesome-asgi] for a listing of some of the others.

[ASGI Specification]: https://asgi.readthedocs.io/en/latest/
[awesome-asgi]: https://github.com/florimondmanca/awesome-asgi
```

> The following are ideal for running your application in a development environment.

Reviewer: ```suggestion
```

### Running from the command line

The simplest way to run `uvicorn` is from the command line.

```bash
$ uvicorn example:app --reload --port 5000
```

> The ASGI application should be specified in the form `path.to.module:instance.path`

Reviewer: What does `instance.path` mean here? As far as I know it's just `path.to.module:asgi_app_global`, right?

Author: I got this referring from either existing documentation or from FastAPI. It was supposed to mean `main:app` or `app.main:app`

Reviewer: Yeah those sound right but `module:instance.path` does not

Author: Agreed. I'll change that!

Use `--reload` to turn on auto-reloading. 
???+ note
    `--reload` is recommended only for local use, during the development phase.
    
    The `--reload` and `--workers` arguments are **mutually exclusive**.

To see the complete set of available options, use `uvicorn --help`:

> <!-- :cli_usage: -->

Reviewer: Can we just link to the `Introduction` section of the docs instead of repeating the CLI options here?

Author: I was thinking we add these here, and link this in introduction.🧐

Reviewer: Let's start with the smallest change possible, we can always make more sweeping changes to the docs later on if it's warranted

                                  stdlib ssl module's)  [default: 0]
  --ssl-ca-certs TEXT             CA certificates file
  --ssl-ciphers TEXT              Ciphers to use (see stdlib ssl module's)
                                  [default: TLSv1]
  --header TEXT                   Specify custom default HTTP response headers
                                  as a Name:Value pair
  --version                       Display the uvicorn version and exit.
  --app-dir TEXT                  Look for APP in the specified directory, by
                                  adding this to the PYTHONPATH. Defaults to
                                  the current working directory.  [default: .]
  --h11-max-incomplete-event-size INTEGER
                                  For h11, the maximum number of bytes to
                                  buffer of an incomplete event.
  --factory                       Treat APP as an application factory, i.e. a
                                  () -> <ASGI app> callable.
  --help                          Show this message and exit.
```

See the [settings documentation](../settings.md) for more details on the supported options for running uvicorn.

> ### Running programmatically

Reviewer: I think we should also give an example of using `uvicorn.Config` + `uvicorn.Server` manually. I know that @Tinche and I think @agronholm  (sorry if I am misremembering) use this style a lot . It lets you take control of the event loop as well as startup/shutdown, configuring logging and several other things.

Author: I haven’t tried that implementation yet. I’ll add that too!

Reviewer 2: We might want to deal with this separately as it could be added to the existing docs -- not specifically related to this "enhanced" guide. Also, I believe we already mentioned this somewhere? There might be an old PR about it.

Reviewer: Ah yes I missed it! https://github.com/encode/uvicorn/blob/master/docs/index.md#running-programmatically

Uvicorn can also be run programmatically. To run directly from within a Python program, you should use `uvicorn.run(app, **config)`. 

**`example.py`**

```python
import uvicorn

class App:
    ...

app = App()

if __name__ == "__main__":
    uvicorn.run("example:app", host="127.0.0.1", port=5000, log_level="info")
```

The set of configuration options available is the same as for the options listed under running from commandline.

!!! note
    Note that the application instance itself *can* be passed instead of the app
    import string.

    ```python
    uvicorn.run(app, host="127.0.0.1", port=5000, log_level="info")
    ```

    However, this style only works if you are not using multiprocessing (`workers=NUM`)
    or reloading (`reload=True`), so we recommend using the import string style.

    Also note that in this case, you should put `uvicorn.run` into `if __name__ == '__main__'` clause in the main module.

>     The `reload` and `workers` parameters are **mutually exclusive**.

Reviewer: ```suggestion
    Remember, the `reload` and `workers` parameters are **mutually exclusive**.
```

# Deployment

Server deployment is a complex area, that will depend on what kind of service you're deploying Uvicorn onto.

### TL;DR
As a general rule, you probably want to:

* Run `uvicorn --reload` from the command line for local development.
* Run `gunicorn -k uvicorn.workers.UvicornWorker` for production (non containerized).
* Run behind Nginx for self-hosted deployments.
* Finally, run everything behind a CDN for caching support, and serious DDOS protection.

## Quick Start

> This section is meant to get you started serving an ASGI application as quick as possible. ASGI applications are applications that adhere to the [ASGI Specification], which is the contract between the ASGI server (Uvicorn) and the ASGI application. Starlette  is just one ASGI framework amongst many, see [awesome-asgi] for a listing of some of the others.

Reviewer 2: I'm not sure why we mention Starlette here, as it is not referred to in the rest of the page. I thought the reference to `awesome-asgi` (thanks) was meant to point to other servers, but it points to other _frameworks_. Yet the documentation is about Uvicorn as a _server_. So maybe it's a bit confusing?

Besides, I think by the point users reach this page, they'll have some understanding of what ASGI is already, don't think we need to explain it again.

So here's my suggestion:

```suggestion
This section is meant to get you started serving an ASGI application as quick as possible.
```

[ASGI Specification]: https://asgi.readthedocs.io/en/latest/
[awesome-asgi]: https://github.com/florimondmanca/awesome-asgi

### Running from the command line

The simplest way to run `uvicorn` is from the command line.

```bash
$ uvicorn example:app --reload --port 5000
```

The ASGI application should be specified in the form `path.to.module:instance.path`

Use `--reload` to turn on auto-reloading. 

> ???+ note

Reviewer 2: It seems `???` makes the admonition collapsible, and `+` makes it un-collapsed by default. Not sure we need it as the admonition is very small. Thoughts on sticking to plain `!!!`, i.e. show this small admonition always?
```suggestion
!!! note
```

    `--reload` is recommended only for local use, during the development phase.
    
    The `--reload` and `--workers` arguments are **mutually exclusive**.

> 

Reviewer 2: ```suggestion
```

# Deployment

Server deployment is a complex area, that will depend on what kind of service you're deploying Uvicorn onto.

### TL;DR
As a general rule, you probably want to:

> * Run `uvicorn --reload` from the command line for local development.

Reviewer 3: ```suggestion
```

as said, reload is a dev feature so I'd remove that if we talk about deployment, on top of that it's already explained in the docs

> * Run `gunicorn -k uvicorn.workers.UvicornWorker` for production (non containerized).

Reviewer 3: why non-containerized ? some people may like to have it thar way and it's a viable option,

> * Run behind Nginx for self-hosted deployments.

Reviewer 3: why for self-hosted deplyments, the use of nginx could be totally justified in a good number of other cases, this doesnt seem right to me

> * Finally, run everything behind a CDN for caching support, and serious DDOS protection.

Reviewer 3: ```suggestion
```
what does everything mean ? I'd remove that, it's too generic, not actionnable

Author: Most of these lines were from current documentation. The deployments.md file.

My initial thought was to replace that with getting-started.md and add section wise files.

See the [settings documentation](../settings.md) for more details on the supported options for running uvicorn.

### Running programmatically

Uvicorn can also be run programmatically. To run directly from within a Python program, you should use `uvicorn.run(app, **config)`. 

**`example.py`**

```python
import uvicorn

class App:
    ...

app = App()

if __name__ == "__main__":
    uvicorn.run("example:app", host="127.0.0.1", port=5000, log_level="info")
```

The set of configuration options available is the same as for the options listed under running from commandline.

!!! note
    Note that the application instance itself *can* be passed instead of the app
    import string.

    ```python
    uvicorn.run(app, host="127.0.0.1", port=5000, log_level="info")
    ```

    However, this style only works if you are not using multiprocessing (`workers=NUM`)
    or reloading (`reload=True`), so we recommend using the import string style.

    Also note that in this case, you should put `uvicorn.run` into `if __name__ == '__main__'` clause in the main module.

    Remember, the `reload` and `workers` parameters are **mutually exclusive**.

> 

Reviewer 3: all this is toatlly redundant with all that part is redundant with https://github.com/encode/uvicorn/blob/master/docs/deployment.md I dont see the interest of having twice the same info disseminated in the docs
